sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +306,103 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
-            try {
-                item.source(insertSourceGen.generateSourceAndCheckConstraints(item.insertValues()));
-            } catch (IOException e) {
-                throw ExceptionsHelper.convertToElastic(e);
+    protected IndexItemResponse insert(ShardUpsertRequest request,
+                                       ShardUpsertRequest.Item item,
+                                       IndexShard indexShard,
+                                       boolean isRetry,
+                                       @Nullable ReturnValueGen returnGen,
+                                       @Nullable InsertSourceGen insertSourceGen) throws Exception {
+        assert insertSourceGen != null : "InsertSourceGen must not be null";
+        BytesReference rawSource;
+        Map<String, Object> source = null;
+        try {
+            if (insertSourceGen instanceof FromRawInsertSource) {
+                rawSource = insertSourceGen.generateSourceAndCheckConstraintsAsBytesReference(item.insertValues());
+            } else {
+                source = insertSourceGen.generateSourceAndCheckConstraints(item.insertValues());
+                rawSource = insertSourceGen.toBytesReference(source);
             }
-        } else {
-            Doc currentDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
-
-            Map<String, Object> updatedSource = updateSourceGen.generateSource(
-                currentDoc,
-                item.updateAssignments(),
-                item.insertValues()
-            );
-
-            if (item.returnValues() != null) {
-                updatedDoc = currentDoc.withUpdatedSource(updatedSource);
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+        item.source(rawSource);
+
+        long version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE ? Versions.MATCH_ANY : Versions.MATCH_DELETED;
+        long seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+        long primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+
+        Engine.IndexResult indexResult = index(item, indexShard, isRetry, seqNo, primaryTerm, version);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            if (source == null) {
+                source = JsonXContent.jsonXContent.createParser(
+                    NamedXContentRegistry.EMPTY,
+                    DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                    BytesReference.toBytes(rawSource)).map();
             }
+            returnvalues = returnGen.generateReturnValues(
+                new Doc(-1,
+                        indexShard.shardId().getIndexName(),
+                        item.id(),
+                        indexResult.getVersion(),
+                        indexResult.getSeqNo(),
+                        primaryTerm,
+                        source,
+                    () -> ""
```

Comment:
The other parameters have a different indentation than the last one. I'm surprised that checkstyle doesn't complain aobut this.

Reply:
I forgot to mention it, the checkstyle complains if i format the lambda aligned to the other values.

Reply:
I think the following passes:

```
diff --git a/sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java b/sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java
index 4001a79418..c9c1212316 100644
--- a/sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java
+++ b/sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java
@@ -347,13 +347,14 @@ public class TransportShardUpsertAction extends TransportShardAction<ShardUpsert
             returnvalues = returnGen.generateReturnValues(
                 // return -1 as docId, the docId can only be retrieved by fetching the inserted document again, which
                 // we want to avoid. The docId is anyway just valid with the lifetime of a searcher and can change afterwards.
-                new Doc(-1,
-                        indexShard.shardId().getIndexName(),
-                        item.id(),
-                        indexResult.getVersion(),
-                        indexResult.getSeqNo(),
-                        primaryTerm,
-                        source,
+                new Doc(
+                    -1,
+                    indexShard.shardId().getIndexName(),
+                    item.id(),
+                    indexResult.getVersion(),
+                    indexResult.getSeqNo(),
+                    primaryTerm,
+                    source,
                     () -> ""
                 )
             );
```